File: lifeos/src/lifeos/_common/subject.py
"""Family-subject marker detection, shared by health + exercise ingestion.

Convention (user-defined): the person an entry belongs to is stated at the
START or END of the message; unmarked text is the user themself.

    "Mi esposa tuvo 121, 79, 61 pulsos"   → subject "esposa", remainder parses
    "108, 72, 66 pulsos de mi esposa"     → subject "esposa"
    "My wife slept 7 hours"              → subject "esposa"

Design: detect + STRIP the marker, then let the existing (unchanged) domain
grammars parse the remainder. The subject is the canonical Spanish relation
word so ES/EN and synonyms ("mujer"/"wife") collapse into one label that the
graph layer can resolve against the hub's typed relation edges.

Proper-name markers ("Ana tuvo …") are deliberately NOT supported in v1:
without a roster of known names the pattern "<Word> tuvo" is far too
false-positive-prone (any capitalized sentence start). Precision-first.

Public:
    detect_subject(text) → SubjectMatch | None
    detect_query_subject(query) → str | None
"""

from __future__ import annotations

import re
from dataclasses import dataclass

# Canonical ES relation label per accepted marker word. Keys are matched
# case-insensitively; accent variants are listed explicitly so dictation
# without accents ("mi mama") still matches.
_RELATION_CANON: dict[str, str] = {
    # ES
    "esposa": "esposa", "mujer": "esposa",
    "esposo": "esposo", "marido": "esposo",
    "mamá": "mamá", "mama": "mamá", "madre": "mamá",
    "papá": "papá", "papa": "papá", "padre": "papá",
    "hijo": "hijo", "hija": "hija",
    "hermano": "hermano", "hermana": "hermana",
    "abuelo": "abuelo", "abuela": "abuela",
    "suegro": "suegro", "suegra": "suegra",
    "tío": "tío", "tio": "tío", "tía": "tía", "tia": "tía",
    "primo": "primo", "prima": "prima",
    "novio": "novio", "novia": "novia",
    # EN → canonical ES label (single vocabulary downstream)
    "wife": "esposa", "husband": "esposo",
    "mom": "mamá", "mother": "mamá",
    "dad": "papá", "father": "papá",
    "son": "hijo", "daughter": "hija",
    "brother": "hermano", "sister": "hermana",
}

_RELATION_ALT = "|".join(
    sorted((re.escape(k) for k in _RELATION_CANON), key=len, reverse=True)
)

# Verbs that may follow a LEADING marker ("Mi esposa TUVO 96…"). They are
# captured separately so callers can retry parsing without the verb when the
# remainder grammar is start-anchored (e.g. the bare blood-pressure triple).
# Third-person ES + simple-past EN forms only — precision over recall.
_VERB_ALT = (
    r"tuvo|tiene|ten[íi]a|trae|anda\s+con|midi[oó]|se\s+midi[oó]|marc[oó]"
    r"|registr[oó]|pes[oó]|se\s+pes[oó]|dijo|durmi[oó]|hizo|se\s+tom[oó]|tom[oó]"
    r"|had|has|got|did|took|measured|weighed|slept|said|is|was"
)

# Leading: "mi esposa [tuvo] …" / "my wife [had] …" — anchored at the start.
_LEADING_RE = re.compile(
    rf"^\s*(?:mi|my)\s+(?P<rel>{_RELATION_ALT})\b"
    rf"(?:\s+(?P<verb>{_VERB_ALT})\b)?"
    r"[\s:,]*",
    re.IGNORECASE,
)

# Trailing: "… de mi esposa" / "… of my wife" — anchored at the end
# (optional closing punctuation tolerated).
_TRAILING_RE = re.compile(
    rf"[\s,;]*\b(?:de|of)\s+(?:mi|my)\s+(?P<rel>{_RELATION_ALT})\s*[.!?]?\s*$",
    re.IGNORECASE,
)

# Query-oriented: a possessive family marker ANYWHERE in a free-form question
# ("¿cómo estaba mi esposa ayer?", "la presión de mi esposa"). Unanchored, so
# it catches mid-sentence markers the start/end-anchored regexes above miss.
# The required "mi|my" possessive is the false-positive guard: a self query
# with no relation word ("mi presión", "resumen de salud") never matches.
_QUERY_RE = re.compile(
    rf"\b(?:mi|my)\s+(?P<rel>{_RELATION_ALT})\b",
    re.IGNORECASE,
)


@dataclass(frozen=True, slots=True)
class SubjectMatch:
    subject: str                     # canonical relation label ("esposa")
    remainder: str                   # text with the marker stripped
    remainder_no_verb: str | None    # leading form with the verb ALSO stripped


# Canonical ES relation label → English relation word, for confirmation copy
# in English installs ("your wife"). Spanish uses the canonical label directly.
_EN_RELATION: dict[str, str] = {
    "esposa": "wife", "esposo": "husband",
    "mamá": "mom", "papá": "dad",
    "hijo": "son", "hija": "daughter",
    "hermano": "brother", "hermana": "sister",
    "abuelo": "grandpa", "abuela": "grandma",
    "suegro": "father-in-law", "suegra": "mother-in-law",
    "tío": "uncle", "tía": "aunt",
    "primo": "cousin", "prima": "cousin",
    "novio": "boyfriend", "novia": "girlfriend",
}


def subject_possessive(subject: str, *, en: bool = False) -> str:
    """Possessive phrasing for a family subject, for chat confirmations.

    ``subject_possessive("esposa")`` → ``"tu esposa"``;
    ``subject_possessive("esposa", en=True)`` → ``"your wife"``. Unknown labels
    are passed through as-is ("tu <label>" / "your <label>").
    """
    rel = (subject or "").strip()
    if en:
        return f"your {_EN_RELATION.get(rel.lower(), rel)}"
    return f"tu {rel}"


def _canon(rel: str) -> str:
    return _RELATION_CANON[rel.lower()]


def detect_subject(text: str) -> SubjectMatch | None:
    """Detect a family-subject marker at the start or end of *text*.

    Returns the canonical subject plus the marker-stripped remainder(s), or
    None when the text is unmarked (→ the entry belongs to the user)."""
    if not text or not isinstance(text, str):
        return None

    m = _LEADING_RE.match(text)
    if m:
        remainder_no_verb = None
        if m.group("verb"):
            # remainder keeps the verb (some grammars key off it, e.g.
            # "slept 7 hours"); remainder_no_verb drops it for start-anchored
            # grammars (e.g. "121, 79, 61 pulsos").
            verb_start = m.start("verb")
            remainder = text[verb_start:].strip()
            remainder_no_verb = text[m.end():].strip()
        else:
            remainder = text[m.end():].strip()
        if remainder or remainder_no_verb:
            return SubjectMatch(
                subject=_canon(m.group("rel")),
                remainder=remainder,
                remainder_no_verb=remainder_no_verb or None,
            )

    m = _TRAILING_RE.search(text)
    if m:
        remainder = text[:m.start()].strip()
        if remainder:
            return SubjectMatch(
                subject=_canon(m.group("rel")),
                remainder=remainder,
                remainder_no_verb=None,
            )
    return None


def detect_query_subject(query: str) -> str | None:
    """Detect the family subject a free-form *query* is about.

    Unlike ``detect_subject`` (which strips a start/end marker from a stored
    entry), this scans the whole query for a possessive family marker
    ("mi esposa", "my wife") in any position and returns the canonical
    relation label ("esposa"), or None when the query is about the user
    themself. Used to decide the recall/search subject; None → self.
    """
    if not query or not isinstance(query, str):
        return None
    m = _QUERY_RE.search(query)
    return _canon(m.group("rel")) if m else None

File: mobile/lib/features/memory/domain/subject.dart
/// Family-subject marker detection (roadmap SLICE A3).
///
/// Ported 1:1 from `lifeos/src/lifeos/_common/subject.py`. Convention
/// (user-defined): the person an entry belongs to is stated at the START or
/// END of the message; unmarked text is the user themself.
///
///   "Mi esposa tuvo 121, 79, 61 pulsos"  -> subject "esposa", remainder parses
///   "108, 72, 66 pulsos de mi esposa"    -> subject "esposa"
///   "My wife slept 7 hours"              -> subject "esposa"
///
/// The subject is the canonical Spanish relation word so ES/EN and synonyms
/// ("mujer"/"wife") collapse into one label the graph layer can resolve against
/// the hub's typed relation edges.
///
/// Proper-name markers ("Ana tuvo ...") are deliberately NOT supported: without
/// a roster of known names the pattern `<Word> tuvo` is far too
/// false-positive-prone. Precision-first — a misdetect must never write a fact
/// against the wrong person.
///
/// Dart port note: Dart's `\b` is ASCII-only, and accented relation words
/// ("mamá", "tío") would break word boundaries. We therefore accent-FOLD the
/// text with a length-preserving 1:1 map ([foldAccents]) before matching, run
/// ASCII-only regexes, and slice the ORIGINAL text at the (index-stable) match
/// offsets to recover the remainder.
library;

/// Canonical ES relation label per accepted marker word. Keys are the
/// accent-FOLDED, lowercased forms (matching runs on folded text); values are
/// the canonical (accented) ES label emitted downstream.
const Map<String, String> _relationCanon = <String, String>{
  // ES
  'esposa': 'esposa', 'mujer': 'esposa',
  'esposo': 'esposo', 'marido': 'esposo',
  'mama': 'mamá', 'madre': 'mamá',
  'papa': 'papá', 'padre': 'papá',
  'hijo': 'hijo', 'hija': 'hija',
  'hermano': 'hermano', 'hermana': 'hermana',
  'abuelo': 'abuelo', 'abuela': 'abuela',
  'suegro': 'suegro', 'suegra': 'suegra',
  'tio': 'tío', 'tia': 'tía',
  'primo': 'primo', 'prima': 'prima',
  'novio': 'novio', 'novia': 'novia',
  // EN -> canonical ES label (single vocabulary downstream)
  'wife': 'esposa', 'husband': 'esposo',
  'mom': 'mamá', 'mother': 'mamá',
  'dad': 'papá', 'father': 'papá',
  'son': 'hijo', 'daughter': 'hija',
  'brother': 'hermano', 'sister': 'hermana',
};

/// Canonical ES relation label -> English relation word, for confirmation copy
/// in English installs ("your wife").
const Map<String, String> _enRelation = <String, String>{
  'esposa': 'wife', 'esposo': 'husband',
  'mamá': 'mom', 'papá': 'dad',
  'hijo': 'son', 'hija': 'daughter',
  'hermano': 'brother', 'hermana': 'sister',
  'abuelo': 'grandpa', 'abuela': 'grandma',
  'suegro': 'father-in-law', 'suegra': 'mother-in-law',
  'tío': 'uncle', 'tía': 'aunt',
  'primo': 'cousin', 'prima': 'cousin',
  'novio': 'boyfriend', 'novia': 'girlfriend',
};

/// Length-preserving accent fold: each accented Latin char -> its ASCII base,
/// 1:1, so match offsets stay valid against the original string.
String foldAccents(String input) {
  const map = <String, String>{
    'á': 'a', 'à': 'a', 'ä': 'a', 'â': 'a',
    'é': 'e', 'è': 'e', 'ë': 'e', 'ê': 'e',
    'í': 'i', 'ì': 'i', 'ï': 'i', 'î': 'i',
    'ó': 'o', 'ò': 'o', 'ö': 'o', 'ô': 'o',
    'ú': 'u', 'ù': 'u', 'ü': 'u', 'û': 'u',
    'ñ': 'n',
    'Á': 'A', 'À': 'A', 'Ä': 'A', 'Â': 'A',
    'É': 'E', 'È': 'E', 'Ë': 'E', 'Ê': 'E',
    'Í': 'I', 'Ì': 'I', 'Ï': 'I', 'Î': 'I',
    'Ó': 'O', 'Ò': 'O', 'Ö': 'O', 'Ô': 'O',
    'Ú': 'U', 'Ù': 'U', 'Ü': 'U', 'Û': 'U',
    'Ñ': 'N',
  };
  final buffer = StringBuffer();
  for (final ch in input.split('')) {
    buffer.write(map[ch] ?? ch);
  }
  return buffer.toString();
}

/// Accent-folded + lowercased form, for case/accent-insensitive comparison.
String normalizeSubject(String s) => foldAccents(s.trim().toLowerCase());

// Relation alternation, folded keys sorted longest-first so "esposa" wins over
// a hypothetical shorter prefix.
final String _relationAlt = (_relationCanon.keys.toList()
      ..sort((a, b) => b.length.compareTo(a.length)))
    .map(RegExp.escape)
    .join('|');

// Third-person ES + simple-past EN verbs that may follow a LEADING marker,
// captured so callers can retry parsing without the verb. Accent-folded forms.
const String _verbAlt = r'tuvo|tiene|tenia|trae|anda\s+con|midio|se\s+midio|marco'
    r'|registro|peso|se\s+peso|dijo|durmio|hizo|se\s+tomo|tomo'
    r'|had|has|got|did|took|measured|weighed|slept|said|is|was';

// Leading: "mi esposa [tuvo] ..." / "my wife [had] ..." — anchored at the start.
final RegExp _leadingRe = RegExp(
  r'^\s*(?:mi|my)\s+(?<rel>' +
      _relationAlt +
      r')\b(?:\s+(?<verb>' +
      _verbAlt +
      r')\b)?[\s:,]*',
  caseSensitive: false,
);

// Trailing: "... de mi esposa" / "... of my wife" — anchored at the end.
final RegExp _trailingRe = RegExp(
  r'[\s,;]*\b(?:de|of)\s+(?:mi|my)\s+(?<rel>' +
      _relationAlt +
      r')\s*[.!?]?\s*$',
  caseSensitive: false,
);

// Query-oriented: a possessive family marker ANYWHERE in a free-form question.
// The required "mi|my" possessive is the false-positive guard.
final RegExp _queryRe = RegExp(
  r'\b(?:mi|my)\s+(?<rel>' + _relationAlt + r')\b',
  caseSensitive: false,
);

// Loose/ingestion-oriented: a possessive family marker ANYWHERE, optionally led
// by a preposition ("de/a/con/para/of/for/to/with mi <rel>"). Used by the
// structured-capture path so a marker that sits neither at the exact start
// ("de mi esposa son 120, 60, 49 pulsos") nor at the exact end ("esto le salió a
// mi papá 135, 89, 95 pulsos") is still attributed. The "mi|my" possessive is
// the false-positive guard — a self reading ("mi presión 120/80") has no
// relation word and never matches.
final RegExp _looseRe = RegExp(
  <String>[
    r'(?:\b(?:de|a|con|para|of|for|to|with)\s+)?\b(?:mi|my)\s+(?<rel>',
    _relationAlt,
    r')\b',
  ].join(),
  caseSensitive: false,
);

/// A detected family-subject marker plus the marker-stripped remainder(s).
class SubjectMatch {
  const SubjectMatch({
    required this.subject,
    required this.remainder,
    this.remainderNoVerb,
  });

  /// Canonical ES relation label ("esposa").
  final String subject;

  /// Original text with the marker stripped.
  final String remainder;

  /// Leading form with the verb ALSO stripped (null when no verb / trailing).
  final String? remainderNoVerb;
}

String _canon(String rel) => _relationCanon[foldAccents(rel.toLowerCase())]!;

/// The folded relation-word alternation (longest-first), shared with adjacent
/// ingestion parsers (e.g. person naming) so there is ONE relation vocabulary.
String get relationAlternation => _relationAlt;

/// Canonical ES relation label for an already accent-folded, lowercased
/// relation word ("papa" → "papá"), or null when it is not a known relation.
String? canonRelation(String foldedRel) => _relationCanon[foldedRel];

/// Detect a family-subject marker at the start or end of [text].
///
/// Returns the canonical subject plus the marker-stripped remainder(s), or
/// null when the text is unmarked (-> the entry belongs to the user).
SubjectMatch? detectSubject(String? text) {
  if (text == null || text.isEmpty) return null;
  final folded = foldAccents(text);

  final lead = _leadingRe.firstMatch(folded);
  if (lead != null) {
    String? remainderNoVerb;
    String remainder;
    final verb = lead.namedGroup('verb');
    if (verb != null && verb.isNotEmpty) {
      // remainder keeps the verb (some grammars key off "slept 7 hours");
      // remainderNoVerb drops it for start-anchored grammars.
      final verbStart = folded.indexOf(verb, lead.start);
      remainder = text.substring(verbStart >= 0 ? verbStart : lead.end).trim();
      remainderNoVerb = text.substring(lead.end).trim();
    } else {
      remainder = text.substring(lead.end).trim();
    }
    if (remainder.isNotEmpty ||
        (remainderNoVerb != null && remainderNoVerb.isNotEmpty)) {
      return SubjectMatch(
        subject: _canon(lead.namedGroup('rel')!),
        remainder: remainder,
        remainderNoVerb:
            (remainderNoVerb != null && remainderNoVerb.isNotEmpty)
                ? remainderNoVerb
                : null,
      );
    }
  }

  final trail = _trailingRe.firstMatch(folded);
  if (trail != null) {
    final remainder = text.substring(0, trail.start).trim();
    if (remainder.isNotEmpty) {
      return SubjectMatch(
        subject: _canon(trail.namedGroup('rel')!),
        remainder: remainder,
      );
    }
  }
  return null;
}

/// Detect the family subject a free-form [query] is about.
///
/// Scans the whole query for a possessive family marker ("mi esposa",
/// "my wife") in any position and returns the canonical relation label, or
/// null when the query is about the user themself (-> self).
String? detectQuerySubject(String? query) {
  if (query == null || query.isEmpty) return null;
  final m = _queryRe.firstMatch(foldAccents(query));
  return m == null ? null : _canon(m.namedGroup('rel')!);
}

/// Loose family-subject detection for the STRUCTURED-CAPTURE ingestion path.
///
/// Unlike [detectSubject] (start/end-anchored, clean marker strip), this catches
/// a possessive family marker ANYWHERE — including a leading `de mi <rel>` or a
/// mid-sentence `a mi <rel>` — and returns the canonical relation label plus the
/// text with just the marker phrase blanked out. The remainder still contains
/// the reading's numbers (the metric parser is un-anchored), so it parses even
/// with leftover filler. Returns null when no possessive family marker is
/// present (→ the entry is the user's own). Precision guard: the "mi|my"
/// possessive is required, so a self reading never matches.
SubjectMatch? detectSubjectLoose(String? text) {
  if (text == null || text.isEmpty) return null;
  final m = _looseRe.firstMatch(foldAccents(text));
  if (m == null) return null;
  final remainder =
      '${text.substring(0, m.start)} ${text.substring(m.end)}'.trim();
  return SubjectMatch(
    subject: _canon(m.namedGroup('rel')!),
    remainder: remainder,
  );
}

/// Possessive phrasing for a family subject, for chat confirmations.
/// `subjectPossessive("esposa")` -> "tu esposa"; with [en] -> "your wife".
String subjectPossessive(String subject, {bool en = false}) {
  final rel = subject.trim();
  if (en) return 'your ${_enRelation[rel.toLowerCase()] ?? rel}';
  return 'tu $rel';
}
